patch/llpcPatchDescriptorLoad.cpp

```diff
@@ -78,6 +78,84 @@ PatchDescriptorLoad::PatchDescriptorLoad()
 
+// =====================================================================================================================
+/// Create a non-strided, non-swizzled buffer descriptor.
+///
+/// @param baseAddr (i64) base address of the descriptor (top 16 bits are ignored);
+///                 for convenience, an <N x i32> is also accepted (using the low 48 bits)
+/// @param size (i32) size of the referenced buffer in bytes
+Value* PatchDescriptorLoad::buildBufferDescriptor(Value* baseAddr,
+                                                  Value* size,
+                                                  Instruction* insertPt)
+{
+    IRBuilder<> builder(insertPt);
+    Value* bufDesc = UndefValue::get(m_pContext->Int32x4Ty());
+
+    Value* elem0;
+    Value* elem1;
+
+    if (baseAddr->getType() == m_pContext->Int64Ty()) {
+        elem0 = builder.CreateTrunc(baseAddr, m_pContext->Int32Ty());
+        elem1 = builder.CreateLShr(baseAddr, 32, "", true);
+        elem1 = builder.CreateTrunc(elem1, m_pContext->Int32Ty());
+    } else {
+        elem0 = builder.CreateExtractElement(baseAddr, (uint64_t)0);
+        elem1 = builder.CreateExtractElement(baseAddr, (uint64_t)1);
+    }
+
+    // DWORD0
+    bufDesc = builder.CreateInsertElement(bufDesc, elem0, (uint64_t)0);
+
+    // DWORD1
+    elem1 = builder.CreateAnd(elem1, builder.getInt32(UINT16_MAX));
+    bufDesc = builder.CreateInsertElement(bufDesc, elem1, (uint64_t)1);
+
+    // DWORD2
+    bufDesc = builder.CreateInsertElement(bufDesc, size, (uint64_t)2);
+
+    // DWORD3
+#if LLPC_BUILD_GFX10
+    const GfxIpVersion gfxIp = m_pContext->GetGfxIpVersion();
+    if (gfxIp.major < 10)
+#endif
+    {
+        SqBufRsrcWord3 sqBufRsrcWord3 = {};
+        sqBufRsrcWord3.bits.DST_SEL_X = BUF_DST_SEL_X;
+        sqBufRsrcWord3.bits.DST_SEL_Y = BUF_DST_SEL_Y;
+        sqBufRsrcWord3.bits.DST_SEL_Z = BUF_DST_SEL_Z;
+        sqBufRsrcWord3.bits.DST_SEL_W = BUF_DST_SEL_W;
+        sqBufRsrcWord3.gfx6.NUM_FORMAT = BUF_NUM_FORMAT_UINT;
+        sqBufRsrcWord3.gfx6.DATA_FORMAT = BUF_DATA_FORMAT_32;
+        LLPC_ASSERT(sqBufRsrcWord3.u32All == 0x24FAC);
+
+        bufDesc = builder.CreateInsertElement(bufDesc, builder.getInt32(sqBufRsrcWord3.u32All),
+                                              (uint64_t)3);
+    }
+#if LLPC_BUILD_GFX10
+    else if (gfxIp.major == 10)
+    {
+        SqBufRsrcWord3 sqBufRsrcWord3 = {};
+        sqBufRsrcWord3.bits.DST_SEL_X = BUF_DST_SEL_X;
+        sqBufRsrcWord3.bits.DST_SEL_Y = BUF_DST_SEL_Y;
+        sqBufRsrcWord3.bits.DST_SEL_Z = BUF_DST_SEL_Z;
+        sqBufRsrcWord3.bits.DST_SEL_W = BUF_DST_SEL_W;
+        sqBufRsrcWord3.gfx10.FORMAT = BUF_FORMAT_32_UINT;
+        sqBufRsrcWord3.gfx10.RESOURCE_LEVEL = 1;
+        sqBufRsrcWord3.gfx10.OOB_SELECT = 2;
+        LLPC_ASSERT(sqBufRsrcWord3.u32All == 0x21014FAC);
+
+        bufDesc = builder.CreateInsertElement(bufDesc, builder.getInt32(sqBufRsrcWord3.u32All),
+                                              (uint64_t)3);
```

Comment:
(uint64_t)3 -> 3
or 3ull

Reply:
I'd rather not make that change. The reason is a bit involved and technical unfortunately: `Create{Insert,Extract}Element` have two overloads; one where the index is a `Value*`, and one where it's an `uint64_t`. Putting a literal `0` as the operand causes an ambiguous overload error.

So how to fix this ambiguous overload error? `0ull` doesn't work for 64-bit builds, because in 64-bit builds, `uint64_t` is an `unsigned long int`, while `0ull` is an unsigned long long int`, so there's still an ambiguous overload. Conversely, `0ul` doesn't work for 32-bit builds. The simplest remaining option is the explicit cast that I've put into the code.

Using the same here for the non-zero operand is for consistency.

I believe a possible alternative would be to define a C++11 user-defined _u64 literal. I'd be fine with that.

Reply:
Yes, 0ull does not work. I mean you can keep uint64_t(0). For other non-zero literals, 1ull, 2ull, 3ull is enough to express what we want.
Changing (uint64_t)0 to uint64_t(0) is just to keep consistent with Tim's codes.